lib/decode.js

```diff
@@ -14,68 +28,146 @@ import {inspect} from './util.js';
+  typedLiteralTable = new Map(),
+  diagnose,
+  appContextMap = new Map(),
+}) {
   if(!(cborldBytes instanceof Uint8Array)) {
     throw new TypeError('"cborldBytes" must be a Uint8Array.');
   }
 
   // 0xd9 == 11011001
   // 110 = CBOR major type 6
   // 11001 = 25, 16-bit tag size (65536 possible values)
-  let index = 0;
-  if(cborldBytes[index++] !== 0xd9) {
+  if(cborldBytes[0] !== CBOR_TAG_FIRST_BYTE) {
     throw new CborldError(
       'ERR_NOT_CBORLD',
       'CBOR-LD must start with a CBOR major type "Tag" header of `0xd9`.');
   }
 
-  // ensure `cborldBytes` represent CBOR-LD
-  if(cborldBytes[index++] !== 0x05) {
-    throw new CborldError(
-      'ERR_NOT_CBORLD', 'CBOR-LD 16-bit tag must start with `0x05`.');
-  }
+  // process payload with new tag
+  if(cborldBytes[1] == CBORLD_TAG_SECOND_BYTE) {
+    // get suffix to decompress
+    const suffix = _getSuffixFromBytesWithVarint({cborldBytes});
 
-  const compressionMode = cborldBytes[index];
-  if(compressionMode === undefined) {
-    throw new CborldError(
-      'ERR_NOT_CBORLD', 'Truncated CBOR-LD 16-bit tag.');
-  }
+    // decompress CBOR-LD
+    const decompressor = new Decompressor({
+      documentLoader,
+      keywordsTable,
+      urlSchemeTable,
+      stringTable,
+      typedLiteralTable
+    });
+    const result = await decompressor.decompress(
+      {compressedBytes: suffix, diagnose});
 
-  if(!(compressionMode === 0 || compressionMode === 1)) {
-    throw new CborldError(
-      'ERR_NOT_CBORLD',
-      `Unsupported CBOR-LD compression mode "${compressionMode}".`);
+    if(diagnose) {
+      diagnose('Diagnostic JSON-LD result:');
+      diagnose(inspect(result, {depth: null, colors: true}));
+    }
+
+    return result;
   }
+  if(cborldBytes[1] == CBORLD_TAG_SECOND_BYTE_LEGACY) {
+    // legacy support - process CBOR with old tag
+    let index = 2;
+    const compressionMode = cborldBytes[index];
+    if(compressionMode === undefined) {
+      throw new CborldError(
+        'ERR_NOT_CBORLD', 'Truncated CBOR-LD 16-bit tag.');
+    }
 
-  index++;
-  const {buffer, byteOffset, length} = cborldBytes;
-  const suffix = new Uint8Array(buffer, byteOffset + index, length - index);
+    if(!(compressionMode === 0 || compressionMode === 1)) {
+      throw new CborldError(
+        'ERR_NOT_CBORLD',
+        `Unsupported CBOR-LD compression mode "${compressionMode}".`);
+    }
 
-  // handle uncompressed CBOR-LD
-  if(compressionMode === 0) {
-    return cborg.decode(suffix, {useMaps: false});
-  }
+    index++;
+    const {buffer, byteOffset, length} = cborldBytes;
+    const suffix = new Uint8Array(buffer, byteOffset + index, length - index);
+
+    // handle uncompressed CBOR-LD
+    if(compressionMode === 0) {
+      return cborg.decode(suffix, {useMaps: false});
+    }
+
+    const stringTableWithAppContext = new Map(STRING_TABLE);
+
+    for(const [key, value] of appContextMap) {
+      stringTableWithAppContext.set(key, value);
+    }
```

Comment:
This still feels a bit chaotic. The encoder side of this cleaned things up nicely by detecting legacy mode and then generating the legacy tables from that. Then it called the compressor using either the passed in tables or the legacy tables. We should be able to do the same thing here -- so all of the upfront work of determining the decompression tables, etc. is done once and then the decompressor is created and called once at the end.

Reply:
So to be clear, the flow should be:

```
// determine suffix, compression mode, generate legacy tables via helper(s)

// run regular decode (compressionMode=0) or run decompressor
```